### PR TITLE
[FM v0.6.0]--Update docs for HPA autoscaling enhancement

### PR DESCRIPTION
**Pull request by Huanli-Meng**
### Motivation

Function Mesh v0.6.0 enhanced HAP auto-scaling ([code PR #450]( https://github.com/streamnative/function-mesh/pull/450)), therefore add docs accordingly.

### Modification

- Function Mesh v0.6.0
  - Function/Source/Sink CRD configurations
    - Add an item "minReplicas"
  - Scaling:
    - Restructure the doc and add more examples
- Legacy releases
  - Update typos
  - change `kunectl` to `kubectl`
  - change `maxReplica` to `maxReplicas`

### CODE REVIEW

**Review by tpiperatgod — DISMISSED** [review]
lgtm

File: docs/functions/function-crd.md [hunk 1]
```diff
@@ -17,8 +17,9 @@ This table lists Pulsar Function configurations.
 | `tenant` | The tenant of a Pulsar Function. |
 | `namespace` | The Pulsar namespace of a Pulsar Function. |
 | `clusterName` | The Pulsar cluster of a Pulsar Function. |
-| `replicas`| The number of instances that you want to run this Pulsar Function. By default, the `replicas` is set to `1`. |
-| `maxReplicas`| The maximum number of Pulsar instances that you want to run for this Pulsar Function. When the value of the `maxReplicas` parameter is greater than the value of `replicas`, it indicates that the Functions controller automatically scales the Pulsar Functions based on the CPU usage. By default, `maxReplicas` is set to 0, which indicates that auto-scaling is disabled. |
+| `replicas`| The number of instances that you want to run this Pulsar Function. If no value is set, the system will set `1` to it. |
```
Comment (teachwrite) [inline docs/functions/function-crd.md:20]:
```suggestion
| `replicas`| The number of instances that you want to run this Pulsar Function. If no value is set, the system will set it to `1`. |
```

File: docs/functions/function-crd.md [hunk 2]
```diff
@@ -17,8 +17,9 @@ This table lists Pulsar Function configurations.
...
+| `minReplicas`| The minimum number of instances that you want to run for this Pulsar function. If no value is set, the system will set `1` to it. When HPA auto-scaling is enabled, the HPA controller scales up / down the Pods based on the values of the `minReplicas` and `maxReplicas` options. The number of the Pods should be greater than the value of the `minReplicas` while be smaller than the value of the `maxReplicas`.  |
```
Comment (teachwrite) [inline docs/functions/function-crd.md:21]:
```suggestion
| `minReplicas`| The minimum number of instances that you want to run for this Pulsar function. If no value is set, the system will set it to `1`. When HPA auto-scaling is enabled, the HPA controller scales the Pods up / down based on the values of the `minReplicas` and `maxReplicas` options. The number of the Pods should be greater than the value of the `minReplicas` and be smaller than the value of the `maxReplicas`.  |
```

File: docs/scaling.md [hunk 3]
```diff
@@ -4,23 +4,13 @@ category: scaling
...
-This document describes how to scale Pods (Pulsar instances) which are used for running functions, sources, and sinks.
-
-## How it works
-
-With Kubernetes [Horizontal Pod Autoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-Pod-autoscale/), Function Mesh supports automatically scaling the number of Pods (Pulsar instances) that are required to run for Pulsar functions, sources, and sinks.
-
-For resources with HPA configured, the HPA controller monitors the resource's Pods to determine if it needs to change the number of Pod replicas. In most cases, where the controller takes the mean of a per-pod metric value, it calculates whether adding or removing replicas would move the current value closer to the target value.
-
-![scaling](./assets/scaling.png)
+This document describes how to manually and automatically scale Pods (Pulsar instances) which are used for running functions, sources, and sinks.
```
Comment (teachwrite) [inline docs/scaling.md:7]:
```suggestion
This document describes how to manually and automatically scale Pods (Pulsar instances), which are used for running functions, sources, and sinks.
```

File: docs/scaling.md [hunk 4]
```diff
@@ -30,11 +20,27 @@ In CRDs, the `replicas` parameter is used to specify the number of Pods (Pulsar
...
+
+With Kubernetes [Horizontal Pod Autoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-Pod-autoscale/), Function Mesh supports automatically scaling the number of Pods (Pulsar instances) that are required to run Pulsar functions, sources, and sinks.
+
+For resources with HPA configured, the HPA controller monitors the resource's Pods to determine if it needs to change the number of Pod replicas. In most cases, where the controller takes the mean of a per-pod metric value, it calculates whether adding or removing replicas would move the current value closer to the target value.
+
+![scaling](./assets/scaling.png)
+
+### Supported auto-scaling metrics
+
+Function Mesh auto-scales the number of Pods based on the CPU usage, memory usage, custom metrics.
```
Comment (teachwrite) [inline docs/scaling.md:35]:
```suggestion
Function Mesh auto-scales the number of Pods based on the CPU usage, memory usage, and custom metrics.
```

File: docs/scaling.md [hunk 5]
```diff
@@ -30,11 +20,27 @@ In CRDs, the `replicas` parameter is used to specify the number of Pods (Pulsar
...
+
+> **Note**
+>
+> If you have configured autoscaling based on the CPU usage, memory usage, or both of them, you do not need to configure autoscaling based on custom metrics defined in Pulsar Functions or connectors and vice versa.
```
Comment (teachwrite) [inline docs/scaling.md:39]:
```suggestion
> If you have configured autoscaling based on the CPU usage and/or memory usage, you do not need to configure autoscaling based on custom metrics defined in Pulsar Functions or connectors and vice versa.
```

File: docs/scaling.md [hunk 6]
```diff
@@ -44,66 +50,98 @@ Function Mesh auto-scales the number of Pods based on the CPU usage, memory usag
...
+  # Other function configs
+```
 
-  2. Apply the configurations.
+#### Configure autoscaling with built-in metrics
 
-      ```bash
-      kubectl apply -f path/to/source-sample.yaml
-      ```
+Function Mesh supports auto-scales the number of Pods based on [supported built-in metrics](#supported-auto-scaling-metrics).
```
Comment (teachwrite) [inline docs/scaling.md:96]:
```suggestion
Function Mesh supports autoscaling the number of Pods based on [supported built-in metrics](#supported-auto-scaling-metrics).
```

File: docs/scaling.md [hunk 7]
```diff
@@ -44,66 +50,98 @@ Function Mesh auto-scales the number of Pods based on the CPU usage, memory usag
...
-  1. Specify the CPU-based autoscaling metric under `pod.builtinAutoscaler` in the Pulsar Functions CRD.
+    ```bash
+    kubectl apply -f path/to/function-sample.yaml
+    ```
+
+#### Configure autoscaling with custom metrics
+
+Function Mesh supports automatically scaling up the number of Pods based on a custom autoscaling metric. 
+
+- This example shows how to auto-scale the number of Pods if 45% CPU is utilized.
```
Comment (teachwrite) [inline docs/scaling.md:142]:
```suggestion
- This example shows how to auto-scale the number of Pods if 45% CPU is used.
```

File: docs/scaling.md [hunk 8]
```diff
@@ -116,31 +154,33 @@ These examples describe how to auto-scale the number of Pods running Pulsar Func
...
 
       ```bash
-      kubectl apply -f path/to/source-sample.yaml
+      kubectl apply -f path/to/function-sample.yaml
       ```
 
-  >**Note**
-  >
-  > If you specify multiple metrics for the HPA to scale on, the HPA controller evaluates each metric, and proposes a new scale based on that metric. The largest of the proposed scales will be used as the new scale.
+- If a large number of resources are utilized suddenly, you can use the `autoScalingBehavior.scaleUp.stabilizationWindowSeconds` option to adjust the value of the stabilization window to tune performance.
```
Comment (teachwrite) [inline docs/scaling.md:181]:
```suggestion
- If a large number of resources are used suddenly, you can use the `autoScalingBehavior.scaleUp.stabilizationWindowSeconds` option to adjust the value of the stabilization window to tune performance.
```

File: docs/scaling.md [hunk 9]
```diff
@@ -149,25 +189,39 @@ These examples describe how to auto-scale the number of Pods running Pulsar Func
...
+              - type: Percent                      --- [4]
+                value: 100                         --- [5]
+                periodSeconds: 15                  --- [6]
+              - type: Pods                         
+                value: 4                           
+                periodSeconds: 15                  
+              selectPolicy: Max                    --- [7]
       ```
 
+     - [1] `scaleUp`: specifies the rules which are used to control scaling behavior while scaling up.
```
Comment (teachwrite) [inline docs/scaling.md:215]:
```suggestion
     - [1] `scaleUp`: specifies the rules that are used to control scaling behavior while scaling up.
```

File: docs/scaling.md [hunk 10]
```diff
@@ -149,25 +189,39 @@ These examples describe how to auto-scale the number of Pods running Pulsar Func
...
+     - [2] `stabilizationWindowSeconds`: indicates the amount of time the HPA controller should consider previous recommendations to prevent flapping of the number of replicas.
+     - [3] `policies`: a list of policies which regulate the amount of scaling. Each item has the following fields:
```
Comment (teachwrite) [inline docs/scaling.md:217]:
```suggestion
     - [3] `policies`: a list of policies that regulate the amount of scaling. Each item has the following fields:
```

File: docs/scaling.md [hunk 11]
```diff
@@ -149,25 +189,39 @@ These examples describe how to auto-scale the number of Pods running Pulsar Func
...
+       - [4] `type`: can be set to the value `Pods` or `Percent`, which indicates the allowed changes in terms of absolute number of pods or percentage of current replicas.
+       - [5] `periodSeconds`: represents the amount of time in seconds for which the rule should hold true.
```
Comment (teachwrite) [inline docs/scaling.md:219]:
```suggestion
       - [5] `periodSeconds`: represents the amount of time in seconds the rule should hold true for.
```

File: versioned_docs/version-0.1.10/scaling.md [hunk 12]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.1.10/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.1.11/scaling.md [hunk 13]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.1.11/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.1.7/scaling.md [hunk 14]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.1.7/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.1.8/scaling.md [hunk 15]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.1.8/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.1.9/scaling.md [hunk 16]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.1.9/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.2.0/scaling.md [hunk 17]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.2.0/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.3.0/scaling.md [hunk 18]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.3.0/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.4.0/scaling.md [hunk 19]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.4.0/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

File: versioned_docs/version-0.5.0/scaling.md [hunk 20]
```diff
@@ -68,7 +68,7 @@ Deploy the metrics server in the cluster. The Metrics server provides metrics th
 
 These examples describe how to auto-scale the number of Pods running Pulsar Functions.
 
-- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplica` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
+- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is utilized.
```
Comment (teachwrite) [inline versioned_docs/version-0.5.0/scaling.md:71]:
```suggestion
- Function Mesh supports automatically scaling up the number of Pods by updating the `maxReplicas` parameter. In this case, the number of Pods is updated to `8` when 80% CPU is used.
```

**Review by teachwrite — APPROVED** [review]
I have a few suggestions, mostly word order corrections. Otherwise, this LGTM.

**Comment by Huanli-Meng** [comment]
@tpiperatgod I've applied the technical and editorial review comments. Could you please help look into it again and feel free to approve it if there are no more comments? Thanks.

**Review by tpiperatgod — APPROVED** [review]

**Review by tpiperatgod — APPROVED** [review]

**Pull request merged**